### PR TITLE
JAVA-1761: Add OSGi descriptors and JAVA-1762: Shading

### PR DESCRIPTION
**Pull request by tolbertam**
Motivation:

To enable the driver to be used in OSGi containers.

Modifications:

Use maven-bundle-plugin to add OSGi information to the generated
MANIFEST.MF.

Add jctools as a dependency instead of using netty-shaded
MpscLinkedAtomicQueue in DefaultWriteCoalescer.  This was required
because netty does not export its shaded jctools packages in its
Export-Package definition.

Change Reflection.buildFromConfig to use the expectedSuperType's
ClassLoader instead of the current Thread's ClassLoader when loading a
class.  This enables the driver to load classes from the core bundle in
OSGi, but may be problematic in other situations so this may be an
incomplete solution.

Add OSGi integration tests.

Make all dependency versions properties so they may be passed to
failsafe plugin for use in OSGi tests.

Result:

The driver now supports OSGi.

### CODE REVIEW

File: core/pom.xml
Comment (tolbertam) [inline core/pom.xml]: TODO: need to move definition to parent pom, remove version here.

File: integration-tests/pom.xml
Comment (tolbertam) [inline integration-tests/pom.xml:121]: Couldn't find an easy way to not duplicate this.  Putting this on surefire plugin enables IntelliJ to automatically pick it up.

File: core/src/main/java/com/datastax/oss/driver/internal/core/channel/DefaultWriteCoalescer.java
Comment (tolbertam) [inline core/src/main/java/com/datastax/oss/driver/internal/core/channel/DefaultWriteCoalescer.java:33]: Needed because netty shade's jctools and shaded packages are not automatically exported by maven-bundle-plugin, so just added jctools as a dependency, although maybe its better to shade it ourselves since its a tiny dependency.

File: core/src/main/java/com/datastax/oss/driver/internal/core/util/Reflection.java
Comment (tolbertam) [inline core/src/main/java/com/datastax/oss/driver/internal/core/util/Reflection.java:49]: With OSGi, `Thread.currentThread().getContextClassLoader()` is able to load the class (i.e. `ExpontentialReconnectionPolicy`) just fine, but its a different class loader than what was used to initially load the super type (i.e. `ReconnectionPolicy`) we are checking against which was loaded with the bundle it belongs to (the driver core bundle).   To work around that, I changed this to pass in the `ClassLoader` of the super type (see line 180, 186), but this will not work for custom implementations that are not part of the core driver bundle.

In the general case, none of this is an issue for users unless they are running in an embedded runtime (like osgi or a web app container) or doing something tricky with classloaders.

I think if we were to provide a programmatic builder option, such as `Session.builder().withClassLoader(..)` that was only used for classes that aren't part of the core driver, that could be a solution.

File: core-shaded/pom.xml
Comment (tolbertam) [inline core-shaded/pom.xml]: This might be a bit suspect/non-ideal, what do you think?  We rely on the java-driver-core module to generate the shaded manifest for us, and then pull it in here.  I found this a better alternative than explicitly defining the `Export-Package` and `Import-Package` sections which is quite complex.

**Comment by tolbertam** [comment]
TODO:

* [x] OSGi doc
* [x] Shaded doc

File: manual/core/shaded_jar/README.md
Comment (tolbertam) [inline manual/core/shaded_jar/README.md:15]: Since it's a completely different artifact, we don't need to exclude netty as a dependency anymore which is nice.

File: manual/core/shaded_jar/README.md
Comment (tolbertam) [inline manual/core/shaded_jar/README.md:39]: I omitted the [Limitations](https://docs.datastax.com/en/developer/java-driver/3.5/manual/shaded_jar/#limitations) section we had in the original docs because:

1. You can override netty settings in reference the config (under `advanced.netty`).  So a user would no longer need to provide their own `NettyOptions` implementation in most cases.
2. The only real impact from what I can tell is if you want to override `Compressor<ByteBuf> compressor()`, `FrameCodec<ByteBuf> frameCode()`, or `NettyOptions nettyOptions()` in `InternalDriverContext`.  I consider that a somewhat advanced topic and probably extremely atypical, so I think having documentation for that may just add more noise/confusion than it does any good.

File: manual/osgi/README.md
Comment (tolbertam) [inline manual/osgi/README.md:1]: The OSGi docs for 3.x provide a FAQ section which is pretty extensive and includes steps for logging, compression and so on.  It also had some framework specific stuff for tycho and linked to an examples library.

I don't think we need to go so far other than to indicate that we support OSGi and the other sections i've included here.   But let me know if you'd like me to add more detail.

File: pom.xml
Comment (olim7t) [inline pom.xml]: Is this only to inject the versions in the `integration-tests` POM?

If so, I'd rather limit it to the strictly minimal set. We did it for every dependency in 3.x, but I find that annoying, especially for an outsider who is navigating the POMs from somewhere else than an IDE (say, on search.maven.org).

I'm ok with capturing things as properties if there's a good reason (another one is if the version is shared across multiple dependencies), but not by default.

File: core/src/main/java/com/datastax/oss/driver/internal/core/util/Reflection.java
Comment (olim7t) [reply core/src/main/java/com/datastax/oss/driver/internal/core/util/Reflection.java:49]: > With OSGi, Thread.currentThread().getContextClassLoader() is able to load the class (i.e. ExpontentialReconnectionPolicy) just fine, but its a different class loader than what was used to initially load the super type (i.e. ReconnectionPolicy) we are checking against which was loaded with the bundle it belongs to (the driver core bundle).

So what's happening then? The two classes are loaded, does the `isAssignableFrom` call fail?

If OSGI is too complicated (and after reading a bit on it, it seems like it is), then I agree with your suggestion of adding the "user classloader" as a builder argument. But if it's not specified, we should keep the current behavior (context class loader, then `Reflection`'s loader), because I believe it's the right thing to do in a JEE container, if the driver is in the boot classpath. I'll run a test in Tomcat to confirm this.

File: core-shaded/pom.xml
Comment (olim7t) [reply core-shaded/pom.xml]: As discussed privately, the relative reference bothers me a bit out of principle: in theory, you should be able to check out a module as an independent Maven project and build it (provided that the parent POM and all upstream modules are in your local repo).

But, all things considered, there's little chance of that happening, so that's the most pragmatic approach here.

File: core/src/main/java/com/datastax/oss/driver/internal/core/channel/DefaultWriteCoalescer.java
Comment (olim7t) [reply core/src/main/java/com/datastax/oss/driver/internal/core/channel/DefaultWriteCoalescer.java:33]: Yeah, I think it would be a good idea to shade it, both in the regular and in the Netty-shaded core.
The package should be exported, in case a client of the shaded core writes their own flusher.

File: pom.xml
Comment (tolbertam) [reply pom.xml]: 👍 i'll change this to only reflect the versions we need for the tests.  Alternatively, I could just hardcode the version numbers in the failsafe configuration as well, but i can see us forgetting to update those when we update dependencies.

There's also [a maven plugin](https://gitlab.com/josh.cain/dependencyversion-maven-plugin) out there that surfaces dependency versions as properties automatically. It doesn't seem too heavily used, but we could try it out, wdyt?

File: core/src/main/java/com/datastax/oss/driver/internal/core/util/Reflection.java
Comment (tolbertam) [reply core/src/main/java/com/datastax/oss/driver/internal/core/util/Reflection.java:49]: exactly, `isAssignable` from fails because the classloaders used between the two classes is different.   I'll revert it back and add a classloader builder option and we can document that detail in the manual somewhere as well.

File: core-shaded/pom.xml
Comment (tolbertam) [reply core-shaded/pom.xml]: The other possible viable alternative is to add the shaded MANIFEST to the core jar somwhere, and then pull that MANIFEST from the jar when we shade and put it into `META-INF/MANIFEST.MF`.  That would work around the problem of relative referencing, but will add a unrelated file to the core jar.  Is that worth doing or would you rather not add that extra file to the core jar?

File: core/src/main/java/com/datastax/oss/driver/internal/core/channel/DefaultWriteCoalescer.java
Comment (tolbertam) [reply core/src/main/java/com/datastax/oss/driver/internal/core/channel/DefaultWriteCoalescer.java:33]: Sounds good, i'll add that as it's pretty easy to do.

File: core/src/main/java/com/datastax/oss/driver/internal/core/util/Reflection.java
Comment (tolbertam) [reply core/src/main/java/com/datastax/oss/driver/internal/core/util/Reflection.java:49]: Hrmm, thinking about this some more, I think you are right.

If you use `Thread.currentThread.getContextClassLoader()` by default, I expect it to work in the general case and also in a JEE container if the driver is in the boot classpath or loaded in the same webapp.

You could do the same thing with OSGi by loading the driver bundle in the boot classpath as well, which I think could be reasonable, but I think it also means that all libraries the driver depends on also have to be loaded in the boot classpath.  That's a little different than a webapp, where you can have a ton of jars loaded in the webapp.

I'll do a little more research on how typical this is in OSGi land, but maybe that is the right approach.

Comment (tolbertam) [reply core/src/main/java/com/datastax/oss/driver/internal/core/util/Reflection.java:49]: Oh, i got something working and it didn't require something crazy like that (well at least it doesn't seem as crazy to me).  Will push up shortly.

Comment (tolbertam) [reply core/src/main/java/com/datastax/oss/driver/internal/core/util/Reflection.java:49]: For now, I just changed the code to use `Class.forName` with no `ClassLoader` input provided to using the calling class' `ClassLoader` (45f9d32), which seems to work fine as well.  However, I think it may not work in the case where the policy comes from another bundle, which may be solved by using `DynamicImport-Package: *` (see: [this](http://billhiggins.us/blog/2011/02/24/using-reflection-in-osgi/).  Playing with that now.

Comment (tolbertam) [reply core/src/main/java/com/datastax/oss/driver/internal/core/util/Reflection.java:49]: Confirmed, `Dynamic-ImportPackage: *` was needed to allow the core driver to import code from other bundles via reflection.  I added that and a test to exercise it (10be2cd).  I think that will solve our problem for OSGi, but we may have more to do with `ClassLoader`, maybe we can attack that in a future ticket, or see if it becomes a problem for users?  We may be ok doing nothing.

File: pom.xml
Comment (tolbertam) [reply pom.xml]: Unfortunately the maven plugin doesn't work with intellij, best not to use it and only have properties for the things referenced.  Will make that change.

File: core/src/main/java/com/datastax/oss/driver/internal/core/util/Reflection.java
Comment (olim7t) [reply core/src/main/java/com/datastax/oss/driver/internal/core/util/Reflection.java:49]: > isAssignableFrom fails because the classloaders used between the two classes is different

I don't understand, this is with a built-in policy implementation right? How can two types that belong to the same JAR get loaded by different classloaders?

Regarding `Dynamic-ImportPackage: *`, as far as I understand this is the nuclear option, so I still see a value in allowing users to pass their classloader to the builder.

File: core-shaded/pom.xml
Comment (olim7t) [reply core-shaded/pom.xml]: What does it take to pull the file from the other JAR, just a different transformer? If it's a similar amount of lines in the POM, I don't mind adding an unrelated file to the core jar. If it's more than that, we can live with the relative reference.

Comment (tolbertam) [reply core-shaded/pom.xml]: It's just as easy, and more straightforward (as doesn't rely on location of generating file in other module).  I can make that change.

File: core/src/main/java/com/datastax/oss/driver/internal/core/util/Reflection.java
Comment (tolbertam) [reply core/src/main/java/com/datastax/oss/driver/internal/core/util/Reflection.java:49]: > How can two types that belong to the same JAR get loaded by different classloaders?

They are loaded by different classloaders because one of them is loaded via reflection using the given `ClassLoader` and the other isn't.   

i.e.:

`ReconnectionPolicy.class` that gets passed to `Reflection.buildFromConfig` is referenced in code in `DefaultDriverContext.buildReconnectionPolicy()`.  Therefore it uses the classloader for the bundle of java-driver-core.

`ExpontentialReconnectionPolicy` gets loaded by the `ClassLoader` via reflection using the `ClassLoader` specified as an input to `Class.forName`.

* `Thread.currentThread().contextClassLoader()` is an `Launcher$AppClassLoader`, its associated with the main thread of the app, which can load classes from bundles.
* `Class.forName` without a `ClassLoader` uses the `ClassLoader` of the caller, `Reflection` (the caller) which was loaded with the bundle, so its the same `ClassLoader` as `ExponentialReconnectionPolicy`, but it does not have access to load code from outside bundles, unless using `DynamicImport-Package: *`.

One solution I just realized is that you could just load the supertype using reflection too using the same `ClassLoader` and then do the `isAssignableFrom` check.  `Thread.currentThread().contexClassLoader()` would actually work for that and may not require dynamic import.  I'll look into that.   That might not be a 100% solution (which is why we may need a way for user to provide ClassLoader), but makes the issue less pressing.

Comment (tolbertam) [reply core/src/main/java/com/datastax/oss/driver/internal/core/util/Reflection.java:49]: Hah, no that doesn't work either, because the resolution of the constructor fails because the `ClassLoader` used to load the class is different than that of the arguments being checked against:

> java.lang.IllegalArgumentException: Expected class ExponentialReconnectionPolicy (specified by advanced.reconnection-policy.class) to have an accessible constructor with arguments (interface com.datastax.oss.driver.api.core.context.DriverContext)

I think the rules we need to work with are:

1. The same `ClassLoader` used to load the super type, (in this case the bundle classloader) must be used to load the input class via reflection.  (So `Class.forName(className)` makes the most sense)

2. The `ClassLoader` loading the class via reflection needs to have access to the class its trying to load.  My current way around this was to use `DynamicImport-Package: *`.

So I really think the best option is to not provide a `ClassLoader` to `Class.forName` (so we use the calling class' `ClassLoader` and use `DynamicImport-Package`.

We could try loading everything in `buildFromConfig` using Reflection (including `DriverContext` for constructor resolution), and even then it may not work and it just feels wrong.  In any case i'm trying it to see what happens.

Comment (tolbertam) [reply core/src/main/java/com/datastax/oss/driver/internal/core/util/Reflection.java:49]: > We could try loading everything in buildFromConfig using Reflection (including DriverContext for constructor resolution), and even then it may not work and it just feels wrong. In any case i'm trying it to see what happens.

That doesn't work.

File: core-shaded/pom.xml
Comment (tolbertam) [reply core-shaded/pom.xml]: Done: 3d32d0017 . Wasn't as easy as I anticipated but no longer relies on relative pathing.

File: core/src/main/java/com/datastax/oss/driver/internal/core/util/Reflection.java
Comment (tolbertam) [reply core/src/main/java/com/datastax/oss/driver/internal/core/util/Reflection.java:49]: I realized that Akka would likely have the same issue and came across [this](https://doc.akka.io/docs/akka/2.5/general/configuration.html#a-word-about-classloaders):

> In several places of the configuration file it is possible to specify the fully-qualified class name of something to be instantiated by Akka. This is done using Java reflection, which in turn uses a ClassLoader. Getting the right one in challenging environments like application containers or OSGi bundles is not always trivial, the current approach of Akka is that each ActorSystem implementation stores the current thread’s context class loader (if available, otherwise just its own loader as in this.getClass.getClassLoader) and uses that for all reflective accesses. This implies that putting Akka on the boot class path will yield NullPointerException from strange places: this is not supported.

I noticed that their manifest also uses `DynamicImport-Package: *` ([sbt config](https://github.com/akka/akka/blob/master/project/OSGi.scala#L36)).

Comment (tolbertam) [reply core/src/main/java/com/datastax/oss/driver/internal/core/util/Reflection.java:49]: Looked more into Akka's class loading behavior today.  Here's how it resolves the `ClassLoader` by default, which matches the documentation from the previous comment:

https://github.com/akka/akka/blob/75f4e2dae297b3e04b309de8709d2dc0abf92be9/akka-actor/src/main/scala/akka/util/Reflect.scala#L162-L178

The gist is, default to `Thread.currentThread().getContextClassLoader()` and if that is null, use the ClassLoader of the caller class (with some exceptions).  In the general case, it will use `getContextClassLoader()`, but that can be overridden when creating the `ActorSystem`, which is exactly what akka-osgi does:

https://github.com/akka/akka/blob/75f4e2dae297b3e04b309de8709d2dc0abf92be9/akka-osgi/src/main/scala/akka/osgi/OsgiActorSystemFactory.scala#L53-L63

It creates a [`BundleDelegatingClassLoader`](https://github.com/akka/akka/blob/75f4e2dae297b3e04b309de8709d2dc0abf92be9/akka-osgi/src/main/scala/akka/osgi/BundleDelegatingClassLoader.scala) that falls back on `ActorSystem`s `ClassLoader`.

This is all very complex, but I think we should probably follow suit on the idea of allowing the user to pass in the `ClassLoader` as a builder option, and otherwise using `Thread.currentThread().getContextClassLoader()` by default.  I don't think we need to go as far as creating a custom OSGi module for using the driver with OSGi, but we can document how to provide a class loader that works.

Comment (tolbertam) [reply core/src/main/java/com/datastax/oss/driver/internal/core/util/Reflection.java:49]: I added `SessionBuilder.withClassLoader(..)` in e0ba792.  I also added a documentation section to our OSGi docs similar to the akka documentation.  I think this is probably the best solution, the default `ClassLoader` resolution is the same as it was from the beginning, and now the user has the option to override it if they need to.  This behavioris also very close to how to akka handles things, short of providing a `driver-osgi` module, which I think might be overkill.

**Comment by tolbertam** [comment]
I think i've addressed everything but let me know if there's anything I missed.   If it's ok, I rebase this against 4.x locally and there were a few changes I needed to account for, but didn't have to change anything fundamental (didn't squash or anything either).  Ok if I force push that up?

**Comment by olim7t** [comment]
👍 to rebase

**Comment by tolbertam** [comment]
> +1 to rebase

Thanks, done!

File: manual/osgi/README.md
Comment (olim7t) [inline manual/osgi/README.md]: Nit: could you respect the 100-character line limit in mardown files? You can show the right margin in IDEA to make this easier, there's also a "wrap to column" plugin.

**Comment by tolbertam** [comment]
Need to do a little work now that jcip annotations dependency is required, will fix that shortly.

File: manual/osgi/README.md
Comment (olim7t) [inline manual/osgi/README.md]: It's also a good idea to wrap long code lines (even if they don't wrap in reality), I think our HTML generated doc doesn't deal very well with them.

File: core/src/main/java/com/datastax/oss/driver/internal/core/util/Reflection.java
Comment (olim7t) [inline core/src/main/java/com/datastax/oss/driver/internal/core/util/Reflection.java:58]: 👍

File: core/src/main/java/com/datastax/oss/driver/api/core/session/SessionBuilder.java
Comment (olim7t) [inline core/src/main/java/com/datastax/oss/driver/api/core/session/SessionBuilder.java]: Maybe add a comment that this is for complex environment like OSGi, and in other case user doesn't need to worry about this.

Comment (olim7t) [reply core/src/main/java/com/datastax/oss/driver/api/core/session/SessionBuilder.java]: Referencing an internal class feels a bit too specific, how about "the classloader that loaded the core driver classes"?

File: integration-tests/src/test/java/com/datastax/oss/driver/osgi/OsgiCustomLoadBalancingPolicyIT.java
Comment (olim7t) [inline integration-tests/src/test/java/com/datastax/oss/driver/osgi/OsgiCustomLoadBalancingPolicyIT.java:25]: 💯 for covering this

File: pom.xml
Comment (olim7t) [inline pom.xml]: I think we don't need to extract this as a property anymore, we don't need it in the OSGi tests since it's shaded.

File: integration-tests/src/test/java/com/datastax/oss/driver/osgi/OsgiBaseIT.java
Comment (olim7t) [inline integration-tests/src/test/java/com/datastax/oss/driver/osgi/OsgiBaseIT.java]: OSGi tests require at least a `mvn package` in the other modules so that the JARs are available. Which is fine given the nature of these tests, but we should mention it here, it's easy to forget and run into errors when you try to execute from the IDE.

**Review by olim7t — APPROVED** [review]
LGTM, glad that we're done with this 🎉 
See a few minor remarks inline.

One last thing: I would *really* like to have the two tickets as separate commits -- 1762 first sounds most logical -- it makes things much clearer when you `git blame` the POMs later. It won't be trivial because everything is intertwined right now. I have a bit of experience doing that so I can take care of it, lmk.

File: manual/osgi/README.md
Comment (tolbertam) [reply manual/osgi/README.md]: Ah, I thought I was.  Apparently my IDE is using a 120 line margin, will fix.

File: pom.xml
Comment (tolbertam) [reply pom.xml]: good catch - will fix.

File: integration-tests/src/test/java/com/datastax/oss/driver/osgi/OsgiBaseIT.java
Comment (tolbertam) [reply integration-tests/src/test/java/com/datastax/oss/driver/osgi/OsgiBaseIT.java]: :+1: will do

**Comment by tolbertam** [comment]
Thanks for reviewing!  I'll process your feedback quickly and push up my changes.

> One last thing: I would really like to have the two tickets as separate commits -- 1762 first sounds most logical -- it makes things much clearer when you git blame the POMs later. It won't be trivial because everything is intertwined right now. I have a bit of experience doing that so I can take care of it, lmk.

Good point.  It actually doesn't seem too bad so I can give that a try if you are ok with that.  My approach would be evaluate each commit and split it into separate commits to divide shading / osgi commits, and then bring everything together in separate commits.

**Comment by olim7t** [comment]
> My approach would be evaluate each commit and split it into separate commits to divide shading / osgi commits, and then bring everything together in separate commits.

Sounds good to me 👍

**Comment by tolbertam** [comment]
Updated to separate changes into commits and tested each commit individually to ensure everything was working.  Then I extracted the java-driver-shaded-guava snapshot out into a separate commit.

Forced pushed, kept the old commits around as [java1761old](https://github.com/datastax/java-driver/tree/java1761old) if needed to look at most recent commit (e8d3ee9).

After we merge and release datastax/java-driver-shaded-guava#1, we can chop off the last commit, retest and this should be good to go!  :crossed_fingers: .

**Comment by tolbertam** [comment]
25.1-jre of java-driver-shaded-guava is now on central, dropped the temporary commit to use the snapshot.  Waiting for CI to run and then should be good to go.  Let me know if see any issues in the two separate commits, if looks good :+1:  to merge.

**Comment by tolbertam** [comment]
CI ran well

**Comment by tolbertam** [comment]
Sorry for the noise, I just happened to notice an issue with the JAVA-1762 commit message:

> java-driver-core-shaded module is not generated as part of the build.

i changed it to

> java-driver-core-shaded module is now generated as part of the build.

**Comment by olim7t** [comment]
Forced-push to address a minor issue in manual (one instance of `ClassLoader` was not in monospace). All set 👍

**Pull request merged**
